nautobot/ipam/models.py

```diff
@@ -425,18 +453,120 @@ def family(self):
             return self.prefix.version
         return None
 
-    def get_duplicates(self):
-        return Prefix.objects.net_equals(self.prefix).filter(vrf=self.vrf).exclude(pk=self.pk)
+    def reparent_subnets(self):
+        """
+        Determine list of child nodes and set the parent to self.
+        """
+        query = Prefix.objects.select_for_update().filter(
+            ~models.Q(id=self.id),  # Don't include yourself...
+            parent_id=self.parent_id,
+            prefix_length__gt=self.prefix_length,
+            ip_version=self.ip_version,
+            network__gte=self.network,
+            broadcast__lte=self.broadcast,
+            # namespace=self.namespace,
+        )
+
+        query.update(parent=self)
+
+    def supernets(self, direct=False, discover_mode=False, for_update=False):
+        """Return supernets of this Prefix."""
+        query = Prefix.objects.all()
+
+        if self.parent is None and not discover_mode:
+            return query.none()
+
+        if discover_mode and direct:
+            raise ValidationError("Direct is incompatible with discover_mode.")
+
+        if for_update:
+            query = query.select_for_update()
+
+        if direct:
+            return query.filter(id=self.parent.id)
+
+        return query.filter(
+            ip_version=self.ip_version,
+            prefix_length__lt=self.prefix_length,
+            network__lte=self.network,
+            broadcast__gte=self.broadcast,
+            # namespace=self.namespace,
+        )
+
+    def subnets(self, direct=False, for_update=False):
+        """Return subnets of this Prefix."""
+        query = Prefix.objects.all()
+
+        if for_update:
+            query = query.select_for_update()
+
+        if direct:
+            return query.filter(parent_id=self.id)
+
+        return query.filter(
+            ip_version=self.ip_version,
+            prefix_length__gt=self.prefix_length,
+            network__gte=self.network,
+            broadcast__lte=self.broadcast,
+            # namespace=self.namespace,
+        )
+
+    def is_child_node(self):
+        """
+        Returns whether I am a child node.
+        """
+        return self.parent is not None
+
+    def is_leaf_node(self):
+        """
+        Returns whether I am leaf node (no children).
+        """
+        return not self.child_prefixes.exists()
+
+    def is_root_node(self):
+        """
+        Returns whether I am a root node (no parent).
+        """
+        return self.parent is None
+
+    def get_ancestors(self, ascending=False):
+        """Return my ancestors."""
+        query = self.supernets()
+        if ascending:
+            query = query.reverse()
+        return query
+
+    def get_children(self):
+        """Return my immediate children."""
+        return self.subnets(direct=True)
```

Comment:
Isn't this redundant with the `child_prefixes` reverse relation?

Reply:
It is. Just had it here for taxonomy completeness for tree methods.

Reply:
> child_prefixes relation actually returns all descendants, not merely direct children.

I don't see how that can be possible - non-child descendants don't have `parent = <this instance>` so how can the reverse relation include them?

Reply:
You are correct. Let's just pretend this didn't happen. So yeah it's redundant at the moment.